torch_geometric/nn/aggr/equilibrium.py

```diff
@@ -0,0 +1,181 @@
+            y = y + momentum
+        return y
+
+
+class EquilibriumAggregation(Aggregation):
+    r"""
+    The graph global pooling layer from the
+    `"Equilibrium Aggregation: Encoding Sets via Optimization"
+    <https://arxiv.org/abs/2202.12795>`_ paper.
+    This output of this layer :math:`\mathbf{y}` is defined implicitly by
+    defining a potential function :math:`F(\mathbf{x}, \mathbf{y})`
+    and regulatization function :math:`R(\mathbf{y})` and the condition
+
+    .. math::
+        \mathbf{y} = \min_\mathbf{y} R(\mathbf{y}) +
+        \sum_{i} F(\mathbf{x}_i, \mathbf{y})
+
+    This implementation use a ResNet Like model for the potential function
+    and a simple L2 norm for the regularizer with learnable weight
+    :math:`\lambda`.
+
+    Args:
+        in_channels (int): The number of channels in the input to the layer.
+        out_channels (float): The number of channels in the ouput.
+        num_layers (List[int): A list of the number of hidden units in the
+            potential function.
+        grad_iter (int): The number of steps to take in the internal gradient
+            descent. (default: :obj:`5`)
+        lamb (float): The initial regularization constant. Is learnable.
+            descent. (default: :obj:`0.1`)
+    """
+    def __init__(self, in_channels: int, out_channels: int,
+                 num_layers: List[int], grad_iter: int = 5, lamb: float = 0.1):
+        super().__init__()
+
+        self.potential = ResNetPotential(in_channels + out_channels, 1,
+                                         num_layers)
+        self.optimizer = MomentumOptimizer()
+        self._initial_lambda = lamb
+        self._labmda = torch.nn.Parameter(Tensor([lamb]), requires_grad=True)
+        self.softplus = torch.nn.Softplus()
+        self.grad_iter = grad_iter
+        self.output_dim = out_channels
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.lamb.data.fill_(self._initial_lambda)
+        reset(self.optimizer)
+        reset(self.potential)
+
+    @property
+    def lamb(self):
+        return self.softplus(self._labmda)
+
+    def init_output(self, index: Optional[Tensor] = None) -> Tensor:
+        index_size = 1 if index is None else int(index.max().item() + 1)
+        return torch.zeros(index_size, self.output_dim,
+                           requires_grad=True).float()
+
+    def reg(self, y: Tensor) -> float:
+        return self.lamb * y.square().mean(dim=-2).sum(dim=0)
+
+    def energy(self, x: Tensor, y: Tensor, index: Optional[Tensor]):
+        return self.potential(x, y, index) + self.reg(y)
+
+    def forward(self, x: Tensor, index: Optional[Tensor] = None, *,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
+        if ptr is not None:
+            raise ValueError(f"{self.__class__} doesn't support `ptr`")
+
+        if dim_size is not None:
```

Comment:
I could technically support this, but I don't know what you'd expect the behavior to be: Unlike sum/mean etc, we can't just assume the input is zero (or if we do, the output would just be random).

Thoughts?

Reply:
Please have a look at https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html. The `dim_size` argument ideally is identical to `index_size = index.max() + 1` in your case. If passed, there is no need to compute it in the first place.

Reply:
I didn't fully get that sorry. If `dim_size` is passed 
- if `dim_size < index_size` I'd expect error,
- if  `dim_size > index_size` I'm not sure what to expect - I guess zero in for those entries?

Reply:
Yes, that is correct.